### PR TITLE
[core] Upgrade saxon to 12.4

### PR DESCRIPTION
**Pull request by jsotuyod**
## Describe the PR

Saxon currently supports versions 10, 11 and 12. How long version 10 will remain to be supported is not clear, making the upgrade now gives us some future-proofing, while allowing us to benefit from further improvements in Saxon.

All changes are to internal classes, so this is backwards compatible.

Saxon 11+ and 12.1+ support Java 8+, so since PMD 7.0.0 there is no incompatibility on this front either.

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 0 violations,<br />
introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/8668351293">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Review by adangel — APPROVED** [review]
Thanks!

I think, we should also list this PR in the release notes under "misc".

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/BaseNodeInfo.java [hunk 1]
```diff
@@ -94,13 +107,20 @@ static AxisIterator iterateList(List<? extends NodeInfo> nodes) {
 
     @SuppressWarnings({"unchecked", "rawtypes"})
```
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/BaseNodeInfo.java:108]: Hm... we can remove raw types, but still need unchecked
([NodeListIterator](https://saxonica.plan.io/projects/saxonmirrorhe/repository/he/revisions/saxon12/entry/src/main/java/net/sf/saxon/tree/iter/NodeListIterator.java) requires exactly `NodeInfo` and not subtypes)

[the remark this answers is not in the document]
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/xpath/internal/BaseNodeInfo.java:108]: I'll try something, but put this in a separate PR...

**Pull request merged**